Makefile

```diff
@@ -1,40 +1,71 @@
-.PHONY: install install-all dev dev-all tests lint docs clean build metadeps updatereqs
+.PHONY: .venv update-pip requirements install install-all dev dev-all tests lint docs doctest doctest2 clean build metadeps updatereqs
 
-install:
-	pip install .
+# Try to autodetect if python3 or python is the python executable used.
+PYTHON := $(shell which python3 2>/dev/null || which python 2>/dev/null)
+PIP := $(shell which pip3 2>/dev/null || which pip 2>/dev/null)
+PYTHONPATH=
+SHELL=/bin/bash
+VENV=.venv
 
-install-all:
-	pip install .[all]
+all:
+ifeq ($(PYTHON),)
+	$(error "No python interpreter found")
+endif
+	@echo Using $(PYTHON)
 
-dev:
-	pip install -e .
+.venv:  ## Set up a new Python virtual environment and install requirements
+	rm -rf .venv/
+	$(PYTHON) -m venv $(VENV)
+	$(MAKE) requirements
 
-dev-all:
-	pip install -e .[all]
+update-pip:  ## Update to latest pip
+	$(PYTHON) -m pip install --upgrade pip
 
-uninstall:
-	pip uninstall quantum-pecos
+requirements: update-pip  ## Install/refresh Python project requirements
+	$(PIP) install --upgrade -r requirements.txt
+	$(PIP) install --upgrade -r docs/requirements.txt
 
-tests: install
-	pytest tests
+install: update-pip  ## Install PECOS
+	$(PIP) install .
+
+install-all: update-pip  ## Install PECOS with all optional dependencies
+	$(PIP) install .[all]
+
+dev: update-pip  ## Install PECOS in editing mode for development
+	$(PIP) install -e .
+
+dev-all: update-pip  ## Install PECOS in editing mode for development with all optional dependencies
+	$(PIP) install -e .[all]
+
+uninstall:  ## Uninstall PECOS
+	$(PIP) uninstall quantum-pecos
 
-lint:
+lint:  ## Run all quality checks / linting
 	pre-commit run --all-files
 
-docs: install
-	pip install -r ./docs/requirements.txt
+docs: install  ## Generate documentation
+	$(PIP) install -r ./docs/requirements.txt
 	cd docs && make clean && make html && cd -
 
-clean:
-	rm -rf *.egg-info dist build docs/_build
+tests: install  ## Run tests
+	pytest tests
+
+doctest:  ## Run doctests
+	sphinx-build -b doctest ./docs ./docs/_build
+
+doctest2:  ## Run doctests using pytest
+	pytest ./docs --doctest-glob=*.rst # --doctest-module
+
+clean:  ## Clean up caches and build artifacts
+	rm -rf *.egg-info dist build docs/_build .pytest_cache/ .ruff_cache/
 
-build: clean
+build: clean  ## Build Python package for upload to pypi
 	python -m build --sdist --wheel -n
```

Comment:
Yeah... I thought about it but didn't address it... alright. should fix that... probably by calling `make metadeps`. I am also seeing it fairly common to have other requirement files for building, testing, and documentation. I am wondering if maybe that is a better pattern in case people don't want to use make and it might call things out better rather needing to look in the makefile... dunno